### DISCUSSION TITLE
Discrepancies in wall clock time.

### DISCUSSION THREAD

**Discussion by srinivasst** [async-profiler/async-profiler#613] [Q&A]
Hi folks, I am trying to identify how long a method ran in my application from the stack traces. My application has many threads (Around 68-70).

I have also added a timer before and after the call to the method in the code.

I see discrepancies between the time reported by the code and the samples observed from the flame graph. 
I am attaching the profiler after starting the Java process. Here is the command I use - `profiler.sh -o collapsed -I '*scheduler/Task.run' -i 1000000000 -t -e wall -f /tmp/st.txt start <PID>`. I then use the pearl script to generate the flame graph. 

According to the code the method ran for around 70 seconds (Measured using `java.lang.System.currentTimeMillis()`) but the samples reported by the flame graph is inconsistent from run to run. Sometimes it is close with 67-68 samples (1 sample = 1 seconds), sometimes it is underreporting by a large % like 50-60 samples. 

The method has both CPU and IO involved.

**Comment by srinivasst** [comment]
I also looked at the code to understand how threads are interrupted and how frames are recorded. If the thread is off CPU - who handles the interrupt for it? Example it is blocked on IO or waiting for a lock. Will some other thread be kicked out to handle the interrupt for the off CPU thread?

**Comment by apangin** [comment]
Sampling is by definition a probabilistic technique.
The higher is the sampling rate - the more accurate is the result.
With 1 second interval (btw, `-i 1s` is a shortcut for `-i 1000000000`) the event rate is very low, so the results may appear inaccurate. Try a shorter interval, e.g. 50ms.

Does the total number of collected samples match the profiling duration?
